milli/src/search/query_tree.rs

```diff
@@ -540,6 +544,29 @@ fn create_query_tree(
     Ok(Operation::or(true, operation_children))
 }
 
+#[derive(Default, Debug)]
+struct MatchingWordCache {
+    all: Vec<Rc<MatchingWord>>,
+    map: HashMap<(String, u8, bool), Rc<MatchingWord>>,
+}
+impl MatchingWordCache {
+    fn insert(&mut self, word: String, typo: u8, prefix: bool) -> Option<Rc<MatchingWord>> {
+        match self.map.entry((word.clone(), typo, prefix)) {
```

Comment:
This relies on the assumption that two `MatchingWord` structures are perfectly equivalent if they were created with the same `(word: String, typo: u8, prefix: bool)` parameters.